src/sentry/db/models/fields/node.py

```diff
@@ -158,45 +191,53 @@ def on_delete(self, instance, **kwargs):
+        # to store it under. Set it to an empty dict and generate a default.
+        if not value:
+            value = {}
+            node_id = self.id_func(value)
+
+        # Case 2: This is a value we've loaded from the database, it should be
+        # decompressed/unpickled, and we should end up with {"node_id": "XYZ"}.
+        # Extract the node_id.
+        elif isinstance(value, six.string_types):
             try:
                 value = pickle.loads(decompress(value))
+                node_id = value.pop('node_id')
+                assert value == {}  # shouldn't be anything else left in there.
+                value = None  # This tells NodeData we need to load the data.
             except Exception as e:
+                # TODO this is a bit dangerous as a failure to read/decode the
+                # node_id will end up with this record being replaced with an
+                # empty value under a new key, potentially orphaning an
+                # original value in nodestore. OTOH if we can't decode the info
+                # here, the node was already effectively orphaned.
                 logger.exception(e)
                 value = {}
-        elif not value:
-            value = {}
+                node_id = self.id_func(value)
 
-        if 'node_id' in value:
-            node_id = value.pop('node_id')
-            data = None
+        # Case 3: A new event body that hasn't been saved into nodestore yet.
+        # It could still have a node_id, that it wants itself to be stored under
+        # so extract that if necessary.
         else:
-            node_id = None
-            data = value
+            node_id = value.pop('node_id', self.id_func(value))
+            if self.wrapper is not None:
+                value = self.wrapper(value)
 
-        if self.wrapper is not None and data is not None:
-            data = self.wrapper(data)
-
-        return NodeData(self, node_id, data)
+        return NodeData(self, node_id, value)
 
     def get_prep_value(self, value):
+        """
+            Prepares the NodeData to be written in a Model.save() call.
+
+            Makes sure the event body is written to nodestore and
+            returns the node_id reference to be written to rowstore.
+        """
         if not value and self.null:
             # save ourselves some storage
             return None
 
-        # We can't put our wrappers into the nodestore, so we need to
-        # ensure that the data is converted into a plain old dict
-        data = value.data
-        if isinstance(data, CANONICAL_TYPES):
-            data = dict(data.items())
-
-        # TODO(dcramer): we should probably do this more intelligently
-        # and manually
-        if not value.id:
-            value.id = nodestore.create(data)
-        else:
-            nodestore.set(value.id, data)
-
+        value.save()
         return compress(pickle.dumps({'node_id': value.id}))
```

Comment:
Why do we pickle/gzip a dict with only one key? Is there a different flow where more data is stored?

Reply:
Doing a little archaeology, it appears that originally this column was used to store a picked+gzipped version of the entire event body. Then when the Nodestore was introduced, we shifted to saving the body in Nodestore, and saving the `node_id` reference to it in this column. But there are still tests to check that the 2 modes of operation are interoperable, so the `compress(pickle(dict))` interface remains the same.

Reply:
Note those are the same tests that are currently failing in travis and I need to rethink my approach slightly to make this work correctly.

Reply:
Interesting. Thank you for the context.